### PR TITLE
[SPARK-33061][SQL] Expose inverse hyperbolic trig functions through sql.functions API

### PR DESCRIPTION
**Pull request by rwpenney**
This patch is a small extension to change-request SPARK-28133, which added inverse hyperbolic functions to the SQL interpreter, but did not include those methods within the Scala `sql.functions._` API. This patch makes `acosh`, `asinh` and `atanh` functions available through the Scala API.

Unit-tests have been added to `sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala`. Manual testing has been done via `spark-shell`, using the following recipe:
```
val df = spark.range(0, 11)
              .toDF("x")
              .withColumn("x", ($"x" - 5) / 2.0)
val hyps = df.withColumn("tanh", tanh($"x"))
             .withColumn("sinh", sinh($"x"))
             .withColumn("cosh", cosh($"x"))
val invhyps = hyps.withColumn("atanh", atanh($"tanh"))
                  .withColumn("asinh", asinh($"sinh"))
                  .withColumn("acosh", acosh($"cosh"))
invhyps.show
```
which produces the following output:
```
+----+--------------------+-------------------+------------------+-------------------+-------------------+------------------+
|   x|                tanh|               sinh|              cosh|              atanh|              asinh|             acosh|
+----+--------------------+-------------------+------------------+-------------------+-------------------+------------------+
|-2.5| -0.9866142981514303|-6.0502044810397875| 6.132289479663686| -2.500000000000001|-2.4999999999999956|               2.5|
|-2.0| -0.9640275800758169| -3.626860407847019|3.7621956910836314|-2.0000000000000004|-1.9999999999999991|               2.0|
|-1.5| -0.9051482536448664|-2.1292794550948173| 2.352409615243247|-1.4999999999999998|-1.4999999999999998|               1.5|
|-1.0| -0.7615941559557649|-1.1752011936438014| 1.543080634815244|               -1.0|               -1.0|               1.0|
|-0.5|-0.46211715726000974|-0.5210953054937474|1.1276259652063807|               -0.5|-0.5000000000000002|0.4999999999999998|
| 0.0|                 0.0|                0.0|               1.0|                0.0|                0.0|               0.0|
| 0.5| 0.46211715726000974| 0.5210953054937474|1.1276259652063807|                0.5|                0.5|0.4999999999999998|
| 1.0|  0.7615941559557649| 1.1752011936438014| 1.543080634815244|                1.0|                1.0|               1.0|
| 1.5|  0.9051482536448664| 2.1292794550948173| 2.352409615243247| 1.4999999999999998|                1.5|               1.5|
| 2.0|  0.9640275800758169|  3.626860407847019|3.7621956910836314| 2.0000000000000004|                2.0|               2.0|
| 2.5|  0.9866142981514303| 6.0502044810397875| 6.132289479663686|  2.500000000000001|                2.5|               2.5|
+----+--------------------+-------------------+------------------+-------------------+-------------------+------------------+
```

### CODE REVIEW

**Comment by zero323** [comment]
This might deserve a separate JIRA ticket. Also could you might want to add a ticket id and `[SQL]` tag to the title i.e. `[SPARK-xxxx][SQL]`.

I also suspect that omitting this in language API might be intentional, as pointed out [here](https://github.com/apache/spark/pull/28593#issuecomment-695861989).

**Comment by rwpenney** [comment]
Thanks @zero323 - I've added tags to the title, as suggested.

I'm sorry, I'm not sure how your link to https://github.com/apache/spark/pull/28593 (which seems to be about timestamp processing) might explain why inverse hyperbolics weren't originally uncluded in `sql.functions._`. Could you clarify, thanks?

**Comment by zero323** [comment]
Sure. I asked there why we provide wrappers around certain functions and not other and  @HyukjinKwon pointed out this:

https://github.com/apache/spark/blob/5af62a2ec74356ce1a97c1371321b3424b674289/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L45-L56

**Comment by rwpenney** [comment]
Thanks for the context.

I agree that hyperbolic trig functions aren't everyone's cup of tea, but I think it would be better to have symmetry: cos & acos, sin & asin, tan & atan are available but we currently only have cosh, sinh & tanh *without* their corresponding inverse functions.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -1427,6 +1427,20 @@ object functions {
    */
   def acos(columnName: String): Column = acos(Column(columnName))
 
+  /**
+   * @return inverse hyperbolic cosine of `e`
+   *
+   * @group math_funcs
+   */
```
Comment (zero323) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:1434]: We should probably have `@since` annotation here and for the remaining ones.

Comment (rwpenney) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:1434]: I wasn't sure whether this is something I should put in during the pull-request, or whether it gets added at a later stage.

Shall I presume that this is destined for Spark-3.1, or maybe 3.0.2?

Comment (zero323) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:1434]: It would be 3.1, as new functions are not added in maintenance releases.

**Comment by HyukjinKwon** [comment]
There have been some discussions about which functions to add. Basically some expressions exist in SparkSQL core just for the sake of other DBMS compatibility. In other language APIs, some expressions make less sense.

It's better to make simplify the call and when to add so it ended up with writing as so at https://github.com/apache/spark/pull/29938#issuecomment-703132270. Let's avoid adding functions there just for the sake of matching.

How often are they used? I don't know enough about that. cc @WeichenXu123 or @srowen. Are they used commonly in daily data science?

**Comment by srowen** [comment]
I imagine these are quite rarely used - more in engineering than anything I can think of in data science. Still for consistency I would not mind adding them, to make languages consistent.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by rwpenney** [comment]
As per my previous comments - I'm not suggesting that these inverse-hypoerboic functions will be widely used, but if `sql.functions` makes ordinary trig-functions available with their inverses, it ought to do the same for hyperbolic trig-functions. If there are concerns about `sql.functions` becoming rather bloated, perhaps at some point we may need a `sql.math.functions` or `sql.functions.math`?

I have, personally, found use-cases where an inverse-sinh is quite handy in data-science applications, for compressing the dynamic range of signed quantities, in the same way `log` is often used for unsigned quantities. Hyperbolic-tangents appear sufficiently frequently in logistic regression models that having an inverse-tanh is also likely to be helpful.

**Comment by zero323** [comment]
> Still for consistency I would not mind adding them, to make languages consistent.

I agree with that, but I wonder if consistency requires `(String) => ...` variants.

**Comment by HyukjinKwon** [comment]
Okay, I am fine with this.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**